include/forward_index_builder.hpp

```diff
@@ -0,0 +1,288 @@
+                for (auto &term_id : *doc_iter) {
+                    term_id = mapping[term_id];
+                }
+            }
+        }
+
+        logger() << "Concatenating batches\n";
+        std::ofstream os(basename);
+        write_header(os, document_count);
+        for (auto batch : enumerate(batch_count)) {
+            std::ifstream is(batch_file(basename, batch));
+            is.ignore(8);
+            os << is.rdbuf();
+        }
+
+        logger() << "Done.\n";
+    }
+
+    void build(std::istream &             is,
+               std::string const &        output_file,
+               read_record_function_type  next_record,
+               process_term_function_type process_term,
+               std::ptrdiff_t             batch_size,
+               std::size_t                threads)
+    {
+        Document_Id    first_document{0};
+        std::ptrdiff_t batch_number = 0;
+
+        std::vector<Record> record_batch;
+        tbb::task_group     batch_group;
+        tbb::concurrent_bounded_queue<int> queue;
+        queue.set_capacity((threads - 1) * 2);
+        while (true) {
+            std::optional<Record> record = std::nullopt;
+            try {
+                if (not (record = next_record(is))) {
+                    auto last_batch_size = record_batch.size();
+                    Batch_Process bp{
+                        batch_number, std::move(record_batch), first_document, output_file};
+                    queue.push(0);
+                    batch_group.run([bp = std::move(bp), process_term, this, &queue]() {
+                        run(std::move(bp), process_term);
+                        int x;
+                        queue.try_pop(x);
+                    });
+                    ++batch_number;
+                    first_document += last_batch_size;
+                    break;
+                }
+            } catch (Warc_Format_Error &err) {
+                continue;
+            }
+            if (record->type() != "response") {
+                continue;
+            }
+            record_batch.push_back(std::move(*record)); // AppleClang is missing value() in Optional
+            if (record_batch.size() == batch_size) {
+                Batch_Process bp{
+                    batch_number, std::move(record_batch), first_document, output_file};
+                queue.push(0);
+                batch_group.run([bp = std::move(bp), process_term, this, &queue]() {
+                    run(std::move(bp), process_term);
+                    int x;
+                    queue.try_pop(x);
+                });
+                ++batch_number;
+                first_document += batch_size;
+                record_batch = std::vector<Record>();
+            }
+        }
+        batch_group.wait();
+        merge(output_file, static_cast<ptrdiff_t>(first_document), batch_number);
+    }
+};
+
+class Plaintext_Record {
```

Comment:
Should Plaintext_Record have url too?

Reply:
All this makes a lot of sense. Thanks.
Let's keep in mind that `Plaintext_Record` is still very important in order to parse collections that dont come in `warc` or `trec` format, like for example a wikipidia dump. 
So please do not remove `Plaintext_Record`

Reply:
Right, but that's why we will need to be careful about defining what we expect of a plaintext format. Anyway, let's do that later.